Feature/reminders

ilikecubesnstuff opened the pull request:
Implemented `&reminder create`, `&reminder delete`, and `&reminder list`.

Commit:
Implementation of reminders.
Added `create`, `delete`, and `list` (using PaginationView).
Updated DB to work in UTC, MemberReminder no longer has `auto_add_now = True`.
Created converter for time durations like "1w3d2h52s", can be reused for `&expire`.
Created `paginated_embed_menus` function to replace the `EmbedGenerator` class.

Commit:
Clean-up imports and code structure.
Isolated long sleeping (>40d) into standalone function.
Cleaned printing and logging statements.
Deleted `embeds.py`, moved all pagination code into `pagination.py`.

Commit:
Bug fixes: timestamp converter and existing reminder scheduling.
`parse_time` uses correct tortoise function now.
`long_sleep_until` typo fixed.
`Reminder.active` deletion no longer throws key errors after bot disconnect/reconnect.

raynichc commented on utils/converters.py:
```diff
@@ -18,4 +23,30 @@ def RegexConverter(arg: str):
+            timestamp = match.group(1)
+        if timestamp < now_timestamp:
+            raise commands.BadArgument("Timestamp cannot correspond to a time in the past.")
+        if timestamp > 8640000000000:  # for some reason this is the maximum??
```
I think it would be better to catch the error from the [fromtimestamp function](https://docs.python.org/3/library/datetime.html#datetime.date.fromtimestamp), instead of using this magic number.

raynichc commented on utils/misc.py:
```diff
@@ -1,3 +1,104 @@
+from tortoise import timezone
+
+
+DEGENERACY_DELAY = timedelta(seconds=1)  # when datetime is in the past, a small amount of time is slept for regardless
```
I don't quite understand why this is necessary, but I don't think I care that much tbh.

ilikecubesnstuff replied:
I put it in to avoid rate-limiting, i.e. when the bot queues all the existing timers at the start, this stops it from sending all those messages at once.

raynichc commented on utils/misc.py:
```diff
@@ -1,3 +1,104 @@
+
+
+def discord_timestamp_string_format(dt: datetime, fmt: str = ''):
+    return f'<t:{int(dt.timestamp())}{f":{fmt}" if fmt else ""}>'
```
I think this is fine, but it does feel like a bit of a mouthful.

raynichc commented on utils/misc.py:
```diff
@@ -1,3 +1,104 @@
+    N = len(s)
+    key = ''
+    value = ''
+    for i, c in enumerate(s):
```
Instead of iterating over the string char by char, I think it would be better to use a regex string such as this `(\d+?)(\D+)`, to identify the components. Iterating over all the matches in the string, then the first group tells you the value and the second group tells you the interval.

raynichc commented on utils/misc.py:
```diff
@@ -1,3 +1,104 @@
+        elif key.startswith('m'):
+            total_delta += timedelta(minutes=value)
+        elif key.startswith('s'):
+            total_delta += timedelta(seconds=value)
```
This can be changed into
```python
codes = {'w': 'weeks', 'd': 'days', 'h': 'hours', 'm': 'minutes', 's': 'seconds'}
total_delta += timedelta(**{codes[key[0]]: value})
```

then the elif would just need to check if key[0] was in the codes' dict keys list.

Commit:
Requested changes: rewritten `utils.misc` functions, renaming and bug fixes.
Rewritten `parse_time` and `discord_timestamp_string_format`.
Renamed `TimestampConverter` to `DatetimeConverter`.
Now queues existing reminders in `on_ready` (so that users are loaded).
Fixed bugs in `DatetimeConverter`.

NathanealV commented on utils/misc.py:
```diff
@@ -1,3 +1,89 @@
+                    month = now.month,
+                    day = now.day,
+                    hour = now.hour,
+                    minute = now.minute,
+                    second = now.second,
+                    microsecond = now.microsecond,
+                    tzinfo = timezone.get_default_timezone(),
+                )
+                total_delta += next - now
+                now = next
+        elif key.startswith('mo'):
+            # python timedelta objects don't work with months
+            # implemented this way to preserve the day of month
+            now = timezone.now() + total_delta
+            if now.day >= 28:           # for when you add 1 month to the 31st of January, and similar
+                now -= ENDMONTH_BUFFER  # subtract a few days so datetime initialization doesn't throw errors
+            for i in range(value):
+                year = now.year
+                month = now.month
+                if month == 12:
+                    month = 0
+                    year += 1
+                next = datetime(
+                    year = year,
+                    month = month + 1,
+                    day = now.day,
+                    hour = now.hour,
+                    minute = now.minute,
+                    second = now.second,
+                    microsecond = now.microsecond,
+                    tzinfo = timezone.get_default_timezone(),
+                )
+                total_delta += next - now
+                now = next
+        else:
+            # remaining cases are handled with "codes" dictionary
+            total_delta += timedelta(**{CODES[key[0]]: value})
+    return total_delta
+
+
```
Might be worth extracting these functions out into a separate util file (e.g., `utils/time.py`)

NathanealV commented on cogs/reminder.py:
```diff
@@ -1,38 +1,145 @@
+
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)  # TODO: Change back to logging.INFO
```
Discussion about the need to set the level of debugging for each file
Can be addressed by setting a `cogs` logger in `bot.py`

NathanealV commented on cogs/reminder.py:
```diff
@@ -1,38 +1,145 @@
+                    task.add_done_callback(self.removal_callback(reminder.reminder_id))
+                    logger.debug(f'Active timer running: {reminder}')
+
+                logger.debug(f'Active: {len(self.active)}')
```
Might be worth changing this to `Active reminders` instead of just `Active` to easily read DEBUG logs.

NathanealV commented on cogs/reminder.py:
```diff
@@ -1,38 +1,145 @@
-        raise NotImplementedError('Command requires implementation and permission set-up.')
+        reminder = await MemberReminder.get_or_none(reminder_id = id)
+        if not reminder:
+            await ctx.send(f'Invalid reminder ID!')
```
This function should return if there is an invalid reminder ID right?

NathanealV reviewed (commented):
Just a couple things from me, otherwise looks solid. Thanks.

Commit:
Improved logging and `util` code.
Moved time-related utility functions to `utils.time`.
Changed to module-name loggers in `updates`, `reminders` and `converters`.
Edited logging messages in `reminders`.
Fixed bug in `&reminder delete`.

NathanealV reviewed (approved):
LGTM!
